mantidproject/mantid
# Fix bug in powder absolute unit normalisation

**Report to:** Ron Smith

**Summary of work**

1. Fixed bug in powder absolute unit normalisation where volume of the sample was actually being got from the vanadium workspace.
2. Have also added handling of errors relating to incorrectly defined sample/vanadium shapes (as described in the original issue).
3. Throw an error (for POLARIS only) if user tries to create a vanadium workspace without both absorption and multiple scattering
4. Because (3) have exposed number of `MSEvents` for multiple scattering correction (as after 3 needed to correct for multiple scattering but with default parameters the algorithm takes a long time). This will also be useful for OSIRIS users.


**To test:**
(0) Copy `POLARISPowderDiffraction.zip` from my babylon `RWaite` and update the paths in `polaris_config_example.yaml`
(1) Test you cannot create a vanadium withoutboth absorption and mult. scatt
```
from isis_powder import polaris, SampleDetails

config_file_path = r"C:\Users\xhg73778\Downloads\POLARISPowderDiffractionFolderSmall\polaris_config_example.yaml"
polaris = polaris.Polaris(config_file=config_file_path, user_name="test", mode="PDF")

polaris.create_vanadium(first_cycle_run_no="125706", multiple_scattering=False, 
                        do_absorb_corrections=False)                  
```
(2) Check that you can't run focussing with `van_normalisation_method="Absolute"` without defining a sample

```
polaris.create_vanadium(first_cycle_run_no="125706", multiple_scattering=True, 
                        do_absorb_corrections=True, mayers_mult_scat_events=1)              
polaris.focus(run_number="125754", input_mode='Individual', do_absorb_corrections=False,
              do_van_normalisation=True, van_normalisation_method="Absolute")

```
(3) Focus with sample defined
```
sample_details = SampleDetails(height=4.0, radius=0.2985, center=[0, 0, 0], shape='cylinder')
sample_details.set_material(chemical_formula='Si', packing_fraction=0.6)
polaris.set_sample_details(sample=sample_details)                      
polaris.focus(run_number="125754", input_mode='Individual', do_absorb_corrections=False,
              do_van_normalisation=True, van_normalisation_method="Absolute")

```

Fixes #35718






---

#### Reviewer ####

Please comment on the points listed below ([full description](http://developer.mantidproject.org/ReviewingAPullRequest.html)).
**Your comments will be used as part of the gatekeeper process, so please comment clearly on what you have checked during your review.** If changes are made to the PR during the review process then your final comment will be the most important for gatekeepers. In this comment you should make it clear why any earlier review is still valid, or confirm that all requested changes have been addressed.

##### Code Review #####

- Is the code of an acceptable quality?
- Does the code conform to the [coding standards](http://developer.mantidproject.org/Standards/)?
- Are the unit tests small and test the class in isolation?
- If there is GUI work does it follow the [GUI standards](http://developer.mantidproject.org/Standards/GUIStandards.html)?
- If there are changes in the release notes then do they describe the changes appropriately?
- Are the release notes saved in a separate file, using Issue or PR number for file name and in the correct location?

##### Functional Tests #####

- Do changes function as described? Add comments below that describe the tests performed?
- Do the changes handle unexpected situations, e.g. bad input?
- Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.

#### Gatekeeper ####

If you need to request changes to a PR then please add a comment and set the review status to "Request changes". This will stop the PR from showing up in the list for other gatekeepers.
